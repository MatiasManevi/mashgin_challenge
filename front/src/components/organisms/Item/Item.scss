@import 'assets/scss/variables';

.item {
	position: relative;
	height: min-content;
	overflow: hidden;

	img {
		border-radius: $item-picture-radius;
		width: $item-picture-w;
		height: $item-picture-h;
		object-fit: cover;
		vertical-align: bottom;
	}

	&:hover {
		cursor: pointer;
		img {
			transform: scale(1.2);
		}
	}

	&::after {
    content: "";
		position: absolute;
		display: block;
		left: 0;
		top: 0;
		width: 100%;
		height: 100%;
		background: rgba(0, 0, 0, 0) linear-gradient(to bottom, rgba(0, 0, 0, 0) 0px, rgba(0, 0, 0, 0.6) 100%) repeat 0 0;
		z-index: 0;
		border-radius: $item-picture-radius;
  }

	.price {
		position: absolute;
		bottom: 10px;
    left: 20px;
		z-index: 1;
	}

	.item__title {
		position: absolute;
		bottom: 10px;
		right: 20px;
		color: $color-white;
		font-size: $font-size-24;
		z-index: 1;
	}
}
