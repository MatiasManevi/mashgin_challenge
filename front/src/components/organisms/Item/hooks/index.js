export { default as useItem } from 'components/organisms/Item/hooks/useItem';
