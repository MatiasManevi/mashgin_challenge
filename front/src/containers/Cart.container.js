import { createContainer } from "unstated-next"
import { useState } from 'react';

export const CartContainer = createContainer(() => {
	const [cartItems, setCartItems] = useState({});

	const addToCart = (item) => {
		let cart = JSON.parse(localStorage.getItem('cart')) || {};
		if (!cart[item.id]) {
			cart[item.id] = item;
		}
		localStorage.setItem('cart', JSON.stringify(cart));
		setCartItems(cart);
	}

	const removeFromCart = (id) => {
		const cart = JSON.parse(localStorage.getItem('cart'));
		if (!cart[id]) return;
		delete cart[id];
		localStorage.setItem('cart', JSON.stringify(cart));
		setCartItems(cart);
	}

	const getCartItems = () => {
		const cart = JSON.parse(localStorage.getItem('cart'));
		if (!cart) return {};
		return Object.values(cart);
	}

	const loadCart = () => {
		const cart = JSON.parse(localStorage.getItem('cart'));
		setCartItems(cart);
	}

	const emptyCart = () => {
		setCartItems({});
		localStorage.removeItem('cart');
	}

	return {
		cartItems,
		getCartItems,
		removeFromCart,
		addToCart,
		loadCart,
		emptyCart
	};
});
