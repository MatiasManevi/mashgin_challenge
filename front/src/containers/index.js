import React from 'react';

import { CartContainer } from './Cart.container';

// The order here is important. Each provider can inherit from the provider before it
const providers = [
	CartContainer.Provider
];

// eslint-disable-next-line react/display-name
export const ContainerProviders = React.memo(({ children }) => {
	// Put the entire app in the lowest part of the chain
	let current = children;

	// Iteratively build the provider contexts (reversed)
	providers.reverse().forEach((Tag) => {
		current = <Tag>{current}</Tag>;
	});

	return current;
});
