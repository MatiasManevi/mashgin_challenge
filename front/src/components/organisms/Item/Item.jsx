import React from 'react';

import './Item.scss';
import { Price, Button } from 'components/atoms';
import { useItem } from 'components/organisms/Item/hooks';

const Item = ({ item }) => {
	const { add, remove, addedToCart } = useItem(item.id);

	return (
		<div className="item">
			<img className="item-picture" src={`/images/${item.image_id}.jpg`} alt={`picture of ${item.name}`} />
			<Price price={item.price} />
			<p className="item__title">{item.name}</p>
			<Button
				color={addedToCart ? "secondary" : "primary"}
				onClick={addedToCart ? () => remove(item.id) : () => add(item)}>
				{addedToCart ?
					<img title="Remove from Cart" src="images/remove-from-cart.png" alt="remove from cart" /> :
					<img title="Add to Cart" src="images/add-to-cart.png" alt="add to cart" />
				}
			</Button>
		</div>
	);
};

export default Item;
