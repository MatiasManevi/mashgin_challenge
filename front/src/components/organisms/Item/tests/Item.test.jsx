import React from 'react';
import { shallow } from 'enzyme';

import Item from '../Item';
import item from './item.fixtures';
import { ContainerProviders } from 'containers';

describe('Item', () => {
	it('should match snapshot', () => {
		const wrapper = shallow(<ContainerProviders><Item item={item} /></ContainerProviders>);
		expect(wrapper).toMatchSnapshot();
	});
});
