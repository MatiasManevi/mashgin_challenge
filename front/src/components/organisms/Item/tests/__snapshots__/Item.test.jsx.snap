// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item should match snapshot 1`] = `
<Provider>
  <Item
    item={
      Object {
        "category_id": 1,
        "id": 1,
        "image_id": "293202f9d9f7f4",
        "name": "Bagel",
        "price": 2,
      }
    }
  />
</Provider>
`;
